// @prerenderer/rollup-plugin/dist/RollupPrerenderPlugin.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const schema_utils_1 = require("schema-utils");
const Options_1 = require("./Options");
const ts_deepmerge_1 = __importDefault(require("ts-deepmerge"));
const path_1 = __importDefault(require("path"));
function RollupPrerenderPlugin(options = {}) {
    (0, schema_utils_1.validate)(Options_1.schema, options, {
        name: 'Prerender Plugin',
        baseDataPath: 'options',
    });
    const opts = ts_deepmerge_1.default.withOptions({ mergeArrays: false }, Options_1.defaultOptions, options);
    return {
        name: 'Prerender Plugin',
        generateBundle: {
            order: 'post',
            handler(output, bundle) {
                return __awaiter(this, void 0, void 0, function* () {
                    const indexPath = opts.indexPath;
                    const entryPath = opts.entryPath || indexPath;
                    const defaultBundle = bundle[entryPath];
                    const { default: Prerenderer } = yield Promise.resolve().then(() => __importStar(require('@prerenderer/prerenderer')));
                    const PrerendererInstance = new Prerenderer(Object.assign({ staticDir: output.dir || '/' }, options));
                    PrerendererInstance.hookServer((server) => {
                        const express = server.getExpressServer();
                        // Express doesn't have complete typings yet
                        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
                        const routes = express._router.stack;
                        routes.forEach((route, i) => {
                            if (route.route && route.route.path === '*') {
                                routes.splice(i, 1);
                            }
                        }, 'post-fallback');
                        express.get('*', (req, res) => {
                            let url = req.path.slice(1, req.path.endsWith('/') ? -1 : undefined);
                            url = url in bundle || url.includes('.') ? url : (url + '/' + indexPath);
                            if (url.startsWith('/')) {
                                url = url.slice(1);
                            }
                            if (options.urlModifier) {
                                url = options.urlModifier(url);
                            }
                            if (url in bundle) {
                                const chunk = bundle[url];
                                if (url.endsWith('.json') && 'source' in chunk) {
                                    const source = chunk.source;
                                    res.json(JSON.parse(typeof source === 'string' ? source : source.toString()));
                                }
                                else {
                                    try {
                                        res.type(path_1.default.extname(url));
                                        res.send('code' in chunk ? chunk.code : chunk.source);
                                    }
                                    catch (e) {
                                        res.status(500);
                                        this.error('Failed to deliver ' + url + ', is the type of the file correct?');
                                    }
                                }
                            }
                            else if (defaultBundle && 'source' in defaultBundle) {
                                res.send(defaultBundle.source);
                            }
                            else {
                                this.error(url + ' not found during prerender');
                                res.status(404);
                            }
                        });
                    });
                    try {
                        yield PrerendererInstance.initialize();
                        const renderedRoutes = yield PrerendererInstance.renderRoutes([...new Set(opts.routes || [])]);
                        // Calculate outputPath if it hasn't been set already.
                        renderedRoutes.forEach(processedRoute => {
                            // Create dirs and write prerendered files.
                            if (!processedRoute.outputPath) {
                                processedRoute.outputPath = path_1.default.join(processedRoute.route, indexPath);
                                if (processedRoute.outputPath.startsWith('/') || processedRoute.outputPath.startsWith('\\')) {
                                    processedRoute.outputPath = processedRoute.outputPath.slice(1);
                                }
                            }
                            if (processedRoute.outputPath === entryPath) {
                                if (options.fallback) {
                                    const fallback = typeof options.fallback === 'string' ? options.fallback : '_fallback';
                                    const ext = path_1.default.extname(processedRoute.outputPath);
                                    const fileName = processedRoute.outputPath.slice(0, -ext.length) + fallback + ext;
                                    if (!(fileName in bundle[processedRoute.outputPath])) {
                                        this.emitFile(Object.assign(Object.assign({}, bundle[processedRoute.outputPath]), { fileName }));
                                    }
                                }
                                delete bundle[processedRoute.outputPath];
                            }
                            const htmlFile = {
                                type: 'asset',
                                source: processedRoute.html.trim(),
                                name: 'Prerendered route ' + processedRoute.route,
                                fileName: processedRoute.outputPath,
                                needsCodeReference: false,
                            };
                            this.emitFile(htmlFile);
                        });
                    }
                    catch (err) {
                        this.warn('Unable to prerender all routes!');
                        if (err instanceof Error) {
                            this.error(err.message);
                        }
                        else if (typeof err === 'object' && err) {
                            this.error(JSON.stringify(err));
                        }
                    }
                    yield PrerendererInstance.destroy();
                });
            },
        },
    };
}
exports.default = RollupPrerenderPlugin;
